src/server/actions/coder.actions.ts

```diff
@@ -0,0 +1,230 @@
+'use server'
+
+import { Action, z } from './action'
+
+// Private helper method to generate username from email and userId
+function generateUsername(email: string, userId: string) {
+    // Extract the part before @ in email
+    const emailUsername = email.split('@')[0] || ''
+
+    // Remove all non-alphanumeric characters and convert to lowercase
+    let username = emailUsername.replaceAll(/[^a-z0-9]/gi, '').toLowerCase()
+
+    // If username is empty or too short, use userId
+    if (username.length === 0) {
+        username = userId.replaceAll(/[^a-z0-9]/gi, '').toLowerCase()
+    }
+
+    // Truncate to less than 10 characters
+    if (username.length >= 32) {
+        username = username.substring(0, 31)
+    }
+
+    return username
+}
```

Comment:
I'm not sure if we should do this vs just using a random id, this seems like it might have dups.